### ISSUE TITLE
MacOS: "QLocalSocket::connectToServer: Invalid name"

### ISSUE THREAD

**Issue by gwerbin** [equalsraf/neovim-qt#936]
Is it possible to add the ability for Neovim-Qt to connect to a Neovim instance that's listening on a filesystem or domain socket?

Currently, if I try to use `--server /path/to/socket` or `--server @sockname` you get an error saying that the listening address is invalid:

```bash
nvim -i NONE -u NONE --headless --listen ./nvim.sock &
nvim-qt --server ./nvim.sock

nvim -i NONE -u NONE --headless --listen @nvim &
nvim-qt --server @nvim
```

This error happens on Neovim-Qt v0.2.16.1 with Neovim v0.5.0.

**Comment by jgehrig** [comment]
See: https://github.com/equalsraf/neovim-qt/wiki/Remote-attachment

Here is a basic example:
```
export NVIM_LISTEN_ADDRESS=127.0.0.1:1234
nvim --headless ~/.vimrc &
nvim-qt --server $NVIM_LISTEN_ADDRESS
```

#734

**Comment by gwerbin** [comment]
@jgehrig sorry if I was unclear, that isn't what I meant. I was asking about using filesystem sockets and/or Unix domain sockets to communicate between Neovim-Qt and the Neovim server. Right now only hostname+port is supported.

**Comment by jgehrig** [comment]
@gwerbin

Yes, it was only an example, `nvr` uses filesystem socket I think.

I think you command `nvim -i NONE -u NONE --headless --listen ./nvim.sock` is the problem. On my Linux machine, the file is not created. I'll dig up a filesystem socket example, that case should be supported and working.

**Comment by jgehrig** [comment]
Give these steps a try:
```
$ export NVIM_LISTEN_ADDRESS=/tmp/nvimsocket
$ nvim --headless --server $NVIM_LISTEN_ADDRESS
$ nvim-qt --server $NVIM_LISTEN_ADDRESS
```

Does this work for you? This is the mechanism `nvr` uses...

**Comment by gwerbin** [comment]
That isn't working either, so something else must be wrong in my setup. I will try to isolate the issue.

**Comment by jgehrig** [comment]
@gwerbin

Re-opening with an updated title since this appears to be MacOS specific?

I have limited access to Mac machines, so it is tough for me to debug issues like this. I am not aware of any MacOS-specific for `QLocalSocket`. I don't see anything obvious in the [Qt QLocalSocket Documentation](https://doc.qt.io/qt-5/qlocalsocket.html#details).

Does `nvr` work for you with local sockets on MacOS? That might be a good starting point for debugging...

**Comment by gwerbin** [comment]
I don't mind learning how to use GDB, but any pointers on getting it set up would be appreciated. Would I need to use a different `-D` option to CMake to create a debug build?

**Comment by jgehrig** [comment]
My ability to use a Macs is embarrassingly limited :sweat_smile: 

With that said... I'm happy to help in any way I can!

> any pointers on getting it set up

I remember setting up GDB on MacOS being difficult. MacOS has strict permissions and getting GDB whitelisted so it has the correct level of access is not straightforward.

I'd recommend using whatever debugger MacOS allows (X-Code? LLDB?). Look for an "Attach to process..." option.

> Would I need to use a different -D option to CMake to create a debug build?

For the debugger, no. For `qDebug`, yes. Sometimes print debugging is the easiest way to debug...

You can do a build with `-DCMAKE_BUILD_TYPE=Debug`, and instrument the areas of interest in the code with `qDebug()` statements. You can then run the scenario and guess what happened.

Hacky, but sometimes it is the quickest way to solve the problem. Getting the debugger configured isn't always easy.

**Comment by gwerbin** [comment]
Configuration was easy! LLDB "just worked" (once I typed in my login password).

However, it looks like I'm looking at raw assembly code, which is beyond me 😆 

```
% lldb $(where nvim-qt)            jobs:1
(lldb) target create "/usr/local/bin/nvim-qt"
Current executable set to '/usr/local/bin/nvim-qt' (x86_64).
(lldb) process launch -s -- --nofork
Process 96507 launched: '/usr/local/bin/nvim-qt' (x86_64)
Process 96507 stopped
* thread #1, stop reason = signal SIGSTOP
    frame #0: 0x0000000100495000 dyld`_dyld_start
dyld`_dyld_start:
->  0x100495000 <+0>: popq   %rdi
    0x100495001 <+1>: pushq  $0x0
    0x100495003 <+3>: movq   %rsp, %rbp
    0x100495006 <+6>: andq   $-0x10, %rsp
Target 0: (nvim-qt) stopped.
```

I'll try rebuilding with `qDebug()`. However, the compile time is fairly long with a clean CMake run; is there some kind of incremental compilation I can do?

**Comment by gwerbin** [comment]
Progress!

I made a debug build and got LLDB running again:

```
$ nvim --headless --listen nvim.sock &
[1] 41393

$ ls nvim.sock
nvim.sock

$ lldb ./bin/neovim-qt
(lldb) b NeovimConnector::connectToSocket
(lldb) process launch -- --nofork --server nvim.sock
(lldb) print path.toUtf8().data()
(char *) $11 = 0x0000000101b3a148 "nvim.sock"
```

But I'm not sure what else to look for at this point.

**Comment by gwerbin** [comment]
To be clear: NVR is _not_ spawning Neovim-Qt here. I was just showing that NVR can connect to the socket without a problem, while `QLocalScket` in Neovim-Qt complains that the address is invalid.

Re: debug symbols, I ran CMake with `-DCMAKE_BUILD_TYPE=Debug` and it looks like LLDB had no problems finding all the variable names and exact locations in the source code.

**Comment by gwerbin** [comment]
Well, this is weird:

* Run `nvim` and call `serverlist()` with no arguments to get the default socket. Neovim-Qt **can** connect.
* Run `nvim` and call `serverstart('./nvim.sock')`, i.e. with a relative path. A socket file `nvim.sock` is created and I can see it. Neovim-Qt **cannot** connect.
* Run `nvim` and call `serverstart($PWD . './nvim.sock')`, i.e. with an absolute path. A socket file `nvim` (_not_ `nvim.sock` !) is created.
  * If I use an absolute path, `nvim-qt --server $PWD/nvim`, Neovim-Qt **can** connect.
  * If I use a relative path, `nvim-qt --server ./nvim`, Neovim-Qt **cannot** connect.

Interestingly, when I use `nvim --headless --listen $PWD/nvim.sock`, Neovim-Qt throws "connection refused" instead of "invalid name". But I can use `nvim --headless +"call serverstart($PWD . '/nvim.sock')"` I can connect with Neovim-Qt fine.

So this is all over the place, and I have no idea how to proceed!

**Comment by jgehrig** [comment]
> Is there a possibility that there's some upstream Neovim issue?

Anything is possible. My guess: this is an unlikely explanation.

> Neovim-Qt throws "connection refused"

Interesting... I wonder if there is some MacOS permissions issue occurring here? Maybe we are being sandboxed?

>- Run nvim and call serverstart($PWD . './nvim.sock'), i.e. with an absolute path. A socket file nvim (not nvim.sock !) is created.
>    - If I use an absolute path, nvim-qt --server $PWD/nvim, Neovim-Qt can connect.
>    - If I use a relative path, nvim-qt --server ./nvim, Neovim-Qt cannot connect.

Can you compare these two cases in the debugger?

Do you observe any differences between the two? Does the error come from Qt or Neovim?

I'd start here. Comparing a working vs non-working example should help narrow where and why the error occurs.

**Comment by gwerbin** [comment]
I'll have to put this on hold for the next few days, but the few things I can respond to:

> My guess: this is an unlikely explanation.

Indeed, but the fact that the `.sock` gets elided from the socket filename is suspicious to me. Maybe I'll take a detour into the Neovim source just to see if that's deliberate or not.

> Does the error come from Qt or Neovim?

The error comes from Qt; it shows up in stderr of the `nvim-qt` process _and_ in the Neovim-Qt GUI window.

**Comment by equalsraf** [comment]
Hello,

I see the same behaviour in Linux

- absolute paths work
- relative paths don't

A debug log shows the following

```
Neovim fatal error "QLocalSocket::connectToServer: Invalid name"
```

I added my own debug statements for good measure,  NeovimConnector::connectToSocket is being called with a relative path as argument. This in turns would call Qt's QLocalsocket::connectToServer

This does not seem to be related to neovim-qt, here is a minimal example with Qt only

```c++
// clang++ -I /usr/include/qt5/ -lQt5Core -lQt5Network  test.cpp
#include <QtNetwork/QLocalSocket>

int main(int argc, char **argv)
{
	QLocalSocket *s = new QLocalSocket();

	s->connectToServer(argv[1]);
       // if this connects it prints "true QLocalSocket::UnknownSocketError"
	qWarning() << s->waitForConnected() << s->error() << s->errorString();
}
```

I can observe 3 distinct cases:

### 1. absolute paths work fine

```
$ file /tmp/nvim9b77JW/0 
/tmp/nvim9b77JW/0: socket
$ ../a.out /tmp/nvim9b77JW/0 
true QLocalSocket::UnknownSocketError "Unknown error"
```

### 2. relative paths with a path component

```
$ file nvim9b77JW/0 
nvim9b77JW/0: socket
$ ./a.out nvim9b77JW/0 
true QLocalSocket::UnknownSocketError "Unknown error"
```

### 3. relative paths in the current folder fail

```
$ file ./0 
./0: socket
$ ../a.out ./0 
false QLocalSocket::ServerNotFoundError "QLocalSocket::connectToServer: Invalid name"
../a.out 0 
false QLocalSocket::ServerNotFoundError "QLocalSocket::connectToServer: Invalid name"
```

I fooled around with a couple of other paths but it does not seem to be related to the path, only to the current directory. I could be wrong though.

I also had a look into qt in https://github.com/qt/qtbase/blob/1daaf613dc56c5ffedefe7ca6938ae5b191f9411/src/network/socket/qlocalsocket_unix.cpp#L299, but no luck so far.

**Comment by equalsraf** [comment]
Before anyone feels tempted to resolve all socket paths into absolute paths, just a reminder that that method also accepts named pipes in windows (\\.\pipe\etc) and likely unix too i.e. don't replace a relative path with an absolute path that does not exist.

**Comment by equalsraf** [comment]
This seems to be the same as seen in this bug https://bugreports.qt.io/browse/QTBUG-50877. From the testing earlier it seems there are two distinct cases, but the upstream issue does not go into them and I was unable to find the source for this.

To verify try this:

1. open an nvim instance and check its value for $NVIM_LISTEN_ADDRESS (it should be inside /tmp/nvimXXXX/0)
2. from any folder in the system call `nvim-qt --server  nvimXXXX/0` (use the path you saw above)

It seems that Qt assumes relative paths passed QLocalSocket are relative to /tmp/. 

The only solution that I can think of is making the path absolute as part of the call (exclude windows for now).

**Comment by equalsraf** [comment]
I opened a PR to fix this, tests are looking good. I'll merge once CI is done.

**Comment by gwerbin** [comment]
Wow, thanks so much for following up with this!
